sdk/storage/Azure.Storage.Files.Shares/tests/DirectoryClientTests.cs

```diff
@@ -1228,6 +1229,235 @@ await foreach (ShareFileItem shareFileItem in directoryClient.GetFilesAndDirecto
             Assert.AreEqual(expectedUri, shareUriBuilder.ToUri());
         }
 
+        [RecordedTest]
+        [ServiceVersion(Min = ShareClientOptions.ServiceVersion.V2021_04_10)]
+        public async Task RenameAsync()
+        {
+            // Arrange
+            await using DisposingShare test = await SharesClientBuilder.GetTestShareAsync();
+            string destDirectoryName = GetNewDirectoryName();
+            ShareDirectoryClient sourceDirectory = InstrumentClient(test.Share.GetDirectoryClient(GetNewDirectoryName()));
+            await sourceDirectory.CreateAsync();
+
+            // Act
+            ShareDirectoryClient destDirectory = await sourceDirectory.RenameAsync(destDirectoryName);
+
+            // Assert
+            await destDirectory.GetPropertiesAsync();
+        }
+
+        [RecordedTest]
+        [ServiceVersion(Min = ShareClientOptions.ServiceVersion.V2021_04_10)]
+        public async Task RenameAsync_DifferentDirectory()
+        {
+            // Arrange
+            await using DisposingShare test = await SharesClientBuilder.GetTestShareAsync();
+            ShareDirectoryClient sourceParentDirectory = InstrumentClient(test.Share.GetDirectoryClient(GetNewDirectoryName()));
+            await sourceParentDirectory.CreateAsync();
+            ShareDirectoryClient sourceDirectory = InstrumentClient(sourceParentDirectory.GetSubdirectoryClient(GetNewDirectoryName()));
+            await sourceDirectory.CreateAsync();
+
+            ShareDirectoryClient destParentDirectory = InstrumentClient(test.Share.GetDirectoryClient(GetNewDirectoryName()));
+            await destParentDirectory.CreateAsync();
+
+            string destDirectoryName = GetNewDirectoryName();
+
+            // Act
+            ShareDirectoryClient destDirectory = await sourceDirectory.RenameAsync(destParentDirectory.Name + "/" + destDirectoryName);
+
+            // Assert
+            await destDirectory.GetPropertiesAsync();
+        }
+
+        [RecordedTest]
+        [ServiceVersion(Min = ShareClientOptions.ServiceVersion.V2021_04_10)]
+        [TestCase(true)]
+        [TestCase(false)]
+        public async Task RenameAsync_ReplaceIfExists(bool replaceIfExists)
```

Comment:
Note that apparently it is legal to overwrite a file with a directory, but not a directory with another directory.